File: tests/roundtrip/bidirectional/consumer/NullableSuspendStructTests.cs
using System;
using System.Threading.Tasks;
using NUnit.Framework;
using roundtrip.nullablesuspendstruct;

public class NullableSuspendStructTests
{
    [Test]
    public async Task NullableStructTaskSignatureAndSynchronousValues()
    {
        Assert.That(typeof(NullableSuspendStructKt).GetMethod("nullableSegment")!.ReturnType,
            Is.EqualTo(typeof(Task<ArraySegment<string>?>)));
        Task<ArraySegment<string?>?> absent = NullableSuspendStructKt.nullableSegment(false);
        Assert.That(await absent, Is.Null);
        var present = await NullableSuspendStructKt.nullableSegment(true);
        Assert.That(present.HasValue, Is.True);
        Assert.That(present.Value.Count, Is.EqualTo(2));
        Assert.That(present.Value[0], Is.Null);
        Assert.That(present.Value[1], Is.EqualTo("value"));
    }

    [Test]
    public async Task NullableStructTaskPreservesSuspendedCompletion()
    {
        foreach (var present in new[] { false, true })
        {
            var gate = new SegmentGate();
            Task<ArraySegment<string?>?> result = NullableSuspendStructKt.delayedSegment(gate);
            Assert.That(result.IsCompleted, Is.False);
            gate.complete(present);
            var value = await result;
            Assert.That(value.HasValue, Is.EqualTo(present));
            if (present) Assert.That(value!.Value[1], Is.EqualTo("resumed"));
        }
    }

    [Test]
    public async Task NullableStructAbstractSlotKeepsNullableTask()
    {
        SegmentSource source = new SegmentSourceImpl();
        Task<ArraySegment<string?>?> absent = source.read(false);
        Assert.That(await absent, Is.Null);
        Assert.That((await source.read(true))!.Value.Count, Is.EqualTo(2));
    }
}

File: tests/roundtrip/consumer/NullableSuspendStructTests.kt
package roundtriptests.nullablesuspendstruct

import NUnit.Framework.TestAttribute
import NUnit.Framework.Legacy.ClassicAssert.AreEqual as assertEquals
import NUnit.Framework.Legacy.ClassicAssert.IsTrue as assertTrue
import System.ArraySegment
import kotlin.coroutines.*
import roundtrip.nullablesuspendstruct.*

private class ConsumerSegmentSource : SegmentSource {
    override suspend fun read(present: Boolean): ArraySegment<String?>? = nullableSegment(present)
}

private class Completion<T> : Continuation<T> {
    var outcome: Result<T>? = null
    override val context: CoroutineContext get() = EmptyCoroutineContext
    override fun resumeWith(result: Result<T>) { outcome = result }
}

private fun <T> start(block: suspend () -> T): Completion<T> {
    val completion = Completion<T>()
    block.startCoroutine(completion)
    return completion
}

class NullableSuspendStructTests {
    @TestAttribute
    fun consumerImplementationKeepsImportedNullableStructSlot() {
        val source: SegmentSource = ConsumerSegmentSource()
        assertTrue(start { source.read(false) }.outcome!!.getOrThrow() == null)
        assertEquals(2, start { source.read(true) }.outcome!!.getOrThrow()!!.Count)
    }

    @TestAttribute
    fun importedNullableStructPreservesNullAndPresent() {
        assertTrue(start { nullableSegment(false) }.outcome!!.getOrThrow() == null)
        assertEquals(2, start { nullableSegment(true) }.outcome!!.getOrThrow()!!.Count)
        val source: SegmentSource = SegmentSourceImpl()
        assertTrue(start { source.read(false) }.outcome!!.getOrThrow() == null)
        assertEquals(2, start { source.read(true) }.outcome!!.getOrThrow()!!.Count)
    }

    @TestAttribute
    fun importedNullableStructResumesThroughSuspendLambda() {
        for (present in listOf(false, true)) {
            val gate = SegmentGate()
            val action: suspend () -> ArraySegment<String?>? = { delayedSegment(gate) }
            val completion = start(action)
            assertTrue(completion.outcome == null)
            gate.complete(present)
            val result = completion.outcome!!.getOrThrow()
            assertEquals(present, result != null)
            if (present) assertEquals(2, result!!.Count)
        }
    }
}
